### DISCUSSION TITLE
Wazuh Integration w/ External Anomaly Detection Module

### DISCUSSION THREAD

**Discussion by dimgiagos44** [wazuh/wazuh#14841] [General]
Hello 👋

After developing an **anomaly detection module** that examines if some amount of traffic contains malicious outliers, i am interested in **integrating** it with the feature-full Wazuh engine. I am envisioning that syslogs forwarded from the Wazuh-Agent would feed my module as well as the Wazuh-Manager, but i havent quite got into the part of **connecting an external module capability with the Wazuh-Manager**. It seems feasible but not too obvious . . .

Any ideas or insights would be extremely useful!

Thanks in advance 🤗

**Comment by dimgiagos44** [comment]
How about creating a custom command for executing a script that wraps my anomaly detection module, and then deploying it as an active-response? 
https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/commands.html
https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/active-response.html

**Comment by CarlosRS9** [comment]
Hello @dimgiagos44 

Let me try to provide you with some information that I hope will be useful.

First of all, just to make sure you are familiar with it, I would like to point out that Wazuh has built-in [Anomaly detection functionality](https://documentation.wazuh.com/current/user-manual/capabilities/anomalies-detection/index.html), thanks to our `rootcheck` and `syscheck` modules. Take a look at our documentation to see if this covers your needs without having to create your own module.

Regarding the solution you propose using [commands](https://documentation.wazuh.com/current/user-manual/capabilities/anomalies-detection/index.html), it is important to keep in mind a couple of factors:

- First, you can indeed make use of our [Active Response](https://documentation.wazuh.com/current/user-manual/capabilities/active-response/index.html) capability to have Wazuh execute your command if it is determined to be necessary. You should note that Wazuh will only execute this command in response to a detected thread, via our Active Response. That is, if no active response is triggered, your command will not be invoked at all.

- Secondly, keep in mind that your module will need to receive certain information as input in order to work. [Command](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/commands.html) allows you to send parameters to your command using the [extra_args](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/commands.html#extra-args) option. Take a look at it to determine if you will be able to send the information your module needs in this way.

Finally, I would like to tell you that, although I have the feeling that it is not going to be compatible with what you are trying to create, there is also the option to integrate your custom module as an external API. Wazuh incorporates the ability to [integrate external APIs](https://documentation.wazuh.com/current/user-manual/manager/manual-integration.html), such as Slack, PagerDuty or in this case custom APIs. Note that the way these integrations work is that Wazuh will send the alerts you want to monitor to your API. If your module provides and listen to an URL then this could be another option to study.

I hope I have not confused you more than necessary and that with these small topics you find the inspiration to achieve what you have proposed.

Regards.

**Reply by dimgiagos44** [reply]
First of all, thanks for the detailed answer @CarlosRS9!

I'm aware of the `rootcheck, syscheck` modules, the thing is that the module i've developed is a ML-based approach to detect some possible anomalies in traffic data.

Concering the commands solution, i have two comments:

- The kind of input my module needs is syslog or traffic generated by an exposed interface (e.g., ne0). I wonder if the module could be configured to read data sent by the Wazuh Agents continuously. Is it possible to forward network/system traffic through the extra_args option?
- I guess an Active Response could be easily triggered at all times if it is configured with an easily-triggered rule..

In the abstract, it seems really feasible to merge Wazuh with the custom module and that's why i am looking into it 😄

**Reply by CarlosRS9** [reply]
I think I now understand better what you are trying to achieve @dimgiagos44.

I guess you could make use of the `Active response` capability, as you mention, since it is indeed possible to configure it to be triggered by low level alerts for certain groups of alerts. However, I do not think that this solution is the most efficient. There are other options that could be better suited for what you want to achieve. Let me guide you by presenting a few concepts and functionalities Wazuh provides that may be useful.


**Log collector capability**

As you may know, Wazuh provides you with the[ log collection](https://documentation.wazuh.com/current/user-manual/capabilities/log-data-collection/index.html) capability, which means that Wazuh is able to process logs from text files, windows event logs or even remote syslog generated by servers and other devides. All those logs are sent to the agent socket by default. However, it is possible to configure Wazuh to sent those logs to [multiple outputs](https://documentation.wazuh.com/current/user-manual/capabilities/log-data-collection/log-data-configuration.html#using-multiple-outputs), which means you can send those to the agent socket and one socket of your choice. Having this in mind, you could build your module to listen to an specific socket and configure Wazuh to feed this socket with the relevant information. You could, for example, monitor syslog files or even the `alerts.json` file itself and target your custom socket.


**Fluentd forwarder**

Another option is the [Fluentd forwarder](https://documentation.wazuh.com/current/user-manual/capabilities/fluent-forwarder.html), which expands on the multiple output topic I mentioned in the previous point. Wazuh allows you to forward messages to a Fluentd server, which is an open source data collector logger. You may find useful information about this topic in our [official documentation](https://documentation.wazuh.com/current/user-manual/capabilities/fluent-forwarder.html) and our [official blog](https://wazuh.com/blog/forward-alerts-with-fluentd/).

If this doesn't get you exactly what you want, maybe it will at least inspire you to find the best way to achieve it.
